[DataAvailability] Extract collection syncing logic from ingestion engine

illia-malachyn opened the pull request:
The `collectionSyncer` type is introduced. Its responsibility is to request/download collections to the local AN's storage.

The logic for the collection syncer is extracted from the ingestion engine. There's a small refactoring to get rid of code duplication. 

Extracted functions and info about what has been changed:
- `lowestHeightWithMissingCollection` -> `findLowestBlockHeightWithMissingCollections` // same
- `updateLastFullBlockReceivedIndex` -> `updateLastFullBlockReceivedIndex` // same
- `requestCollectionsInFinalizedBlock` -> `requestCollections` // immediately/force flag added
- `haveCollection` -> `isCollectionInStorage` // same
- `missingCollectionsAtHeight` -> `findMissingCollectionsAtHeight` // same
- `OnCollection` -> `OnCollectionDownloaded` // same
- `processBackground` -> `RequestCollections` // same
- `checkMissingCollections` -> `requestMissingCollections` // look up logic extracted to `findMissingCollections()`. ingestion engine's unit tests cover this function. (specifically `shouldDownload` statement)
- `requestMissingCollections` -> `downloadMissingCollections` // look up logic extracted to `findMissingCollections()`. ingestion engine's unit tests cover the main request collections loop as well as the logic before 

Closes #7443

peterargue reviewed (commented):
good start. I like that you've made this a component.

It looks like you've done some refactoring. I think we should limit refactoring to only the absolutely necessary (or if there's a compelling reason). That will limit the likelihood of introducing bugs as well as the dev/testing/review overhead.

I'd like to minimize the effort we put into this since it's a fallback and the code is already battle tested.

illia-malachyn commented:
> It looks like you've done some refactoring. I think we should limit refactoring to only the absolutely necessary (or if there's a compelling reason). That will limit the likelihood of introducing bugs as well as the dev/testing/review overhead.

The extracted code is almost the same. I changed some naming and extracted 2 code duplication places into 1 function. Check out the PR's description for more info. I also went through the extracted functions line by line to ensure the logic is identical. I even spent some time debugging it in unit tests because initially I made some mistakes while porting.

peterargue commented on engine/access/ingestion/collection_syncer.go:
```diff
@@ -0,0 +1,389 @@
+	}
+	ready()
+
+	requestCollectionsInterval := time.NewTicker(defaultMissingCollsRequestInterval)
```
we typically use "*Interval" for the actual time.Duration value. how about `requestCollectionsTicker` or `requestCollections`?

peterargue commented on engine/access/ingestion/collection_syncer.go:
```diff
@@ -0,0 +1,389 @@
+	// this means that either we have already received these collections, or the block
+	// may contain unverifiable guarantees (in case this node has just joined the network)
+	if height <= s.lastFullBlockHeight.Value() {
+		s.logger.
+			Info().
```
nit
```suggestion
		s.logger.Info().
```

peterargue reviewed (commented):
I compared the implementations and they are the same besides some light refactoring. added a few small comments, but otherwise looks good.

peterargue commented on engine/access/ingestion/collection_syncer.go:
```diff
@@ -0,0 +1,398 @@
+	return missingCollections, incompleteBlocksCount, nil
+}
+
+// findMissingCollectionsAtHeight returns all missing collections for a specific block height.
+//
+// Expected errors during normal operations:
+// - storage.ErrNotFound if no block is found in storage for the given height
```
since the block is expected to be finalized, the block MUST exist in storage. If it doesn't then it's an exception. Let's rename this to be clearer about the expectation
```suggestion
// findMissingCollectionsAtFinalizedHeight returns all missing collections for a specific finalized block height.
//
// No errors expected during normal operation
```

peterargue commented on engine/access/ingestion/collection_syncer.go:
```diff
@@ -0,0 +1,398 @@
+	for currBlockHeight := firstBlockHeight; currBlockHeight <= lastBlockHeight; currBlockHeight++ {
+		collections, err := s.findMissingCollectionsAtHeight(currBlockHeight)
+		if err != nil {
+			return nil, 0, ErrMissingCollection
```
then this should just return the original error
```suggestion
			return nil, 0, err
```

peterargue commented on engine/access/ingestion/collection_syncer.go:
```diff
@@ -0,0 +1,398 @@
+}
+
+// updateLastFullBlockHeight updates the next highest block height where all previous collections have been indexed.
+// No errors expected during normal operation.
```
let's be consistent with the formatting
```suggestion
//
// No errors expected during normal operation.
```

peterargue commented on engine/access/ingestion/collection_syncer.go:
```diff
@@ -0,0 +1,398 @@
+}
+
+// isCollectionInStorage checks whether the given collection is present in local storage.
+// No errors expected during normal operation
```
```suggestion
//
// No errors expected during normal operation.
```

peterargue commented on engine/access/ingestion/collection_syncer.go:
```diff
@@ -0,0 +1,398 @@
+	// this means that either we have already received these collections, or the block
+	// may contain unverifiable guarantees (in case this node has just joined the network)
+	if height <= s.lastFullBlockHeight.Value() {
+		s.logger.Info().
```
let's downgrade this to debug
```suggestion
		s.logger.Debug().
```

peterargue commented on engine/access/ingestion/collection_syncer.go:
```diff
@@ -0,0 +1,398 @@
+	for currBlockHeight := lastKnownFullBlockHeight + 1; currBlockHeight <= finalizedBlockHeight; currBlockHeight++ {
+		missingCollections, err := s.findMissingCollectionsAtHeight(currBlockHeight)
+		if err != nil {
+			return 0, ErrMissingCollection
```
similar to `findMissingCollections`, this should just return the error
```suggestion
			return 0, err
```

peterargue commented on engine/access/ingestion/collection_syncer.go:
```diff
@@ -0,0 +1,398 @@
+	return newLastFullBlockHeight, nil
+}
+
+// OnCollectionDownloaded indexes and persists a downloaded collection.
+// This is a callback intended to be used with the requester engine.
+func (s *CollectionSyncer) OnCollectionDownloaded(_ flow.Identifier, entity flow.Entity) {
+	collection, ok := entity.(*flow.Collection)
+	if !ok {
+		s.logger.Error().Msgf("invalid entity type (%T)", entity)
+		return
+	}
+
+	err := indexer.IndexCollection(collection, s.collections, s.transactions, s.logger, s.collectionExecutedMetric)
+	if err != nil {
+		s.logger.Error().Err(err).Msg("could not index collection after it has been downloaded")
+		return
+	}
+}
```
This is copied from the original, so no need to make any changes in this PR. I have opened an issue to improve error handling here:
https://github.com/onflow/flow-go/issues/7510
